### PR TITLE
Feature: adding hook display to printRoutes

### PR DESCRIPTION
**Pull request by polymathca**
Re: #2993 

Added ability to include hooks in `printRoutes` output. Can be constrained with an array of hooks in the options object, or will pull the allowed hooks from `lib/hooks` and include any that are defined.

First try at output looks like:

```
   (snip)
   ├── /pages (GET)
   │   • (onTimeout) ["timeoutHook"]
   │   • (onRequest) ["anonymous","authenticate_token"]
   │   ├── / (GET)
   │   │   • (onTimeout) ["timeoutHook"]
   │   │   • (onRequest) ["anonymous","authenticate_token"]
   │   ├── /:id (GET)
   │   │   • (onTimeout) ["timeoutHook"]
   │   │   • (onRequest) ["anonymous","authenticate_token"]
   │   │   └── / (GET)
   │   │       • (onTimeout) ["timeoutHook"]
   │   │       • (onRequest) ["anonymous","authenticate_token"]
   (cont'd)
```

I also exposed the routes array as `fastify.getRoutes()` as it seems like that would be useful for troubleshooting more complex issues (e.g. `console.log( fastify.getRoutes().map(r => r.path) )` to quickly get a list of the routes) but can refactor that if there was a reason that the routes weren't already exposed.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
Do you think it would be possible to augment the find-my-way pretty printer instead? Maybe adding an option `addMetadata` that is a `function` that is called for each route to generate the relevant metadata? In that way we would not have to duplicate code.

**Comment by polymathca** [comment]
For sure. The `route.store` object contains everything we need if we're explicitly passing the `hooksArray` for display. It's set here to default to the `Hooks.supportedHooks` array as before, but can be altered as necessary for debug purposes (e.g. if someone only needs to see `onRequest`)

I have left the `fastify.getRoutes` function in place here, do you see any potential problems with allowing access to that? 

PR: https://github.com/delvedor/find-my-way/pull/185

**Comment by polymathca** [comment]
I will have some time to write tests and documentation this evening (GMT-3)

**Comment by polymathca** [comment]
I just realized this won't pass CI until the changes are merged to `find-my-way`, sorry!

**Comment by mcollina** [comment]
We control the context, so I won't make this overridable right now. Let's go with the hooks and then add things as needed.

**Comment by polymathca** [comment]
> 
> 
> We control the context, so I won't make this overridable right now. Let's go with the hooks and then add things as needed.

Sounds good, I'll revise and push an update!

**Comment by polymathca** [comment]
Okay, I've cleaned everything up. I had forgotten that I started exposing a `getRoutes` function to grab the `routes` object. I have removed that and this PR is strictly reflecting the changes to `fmw` now. Should be ready for review.

One thought re: meta properties. The `context` object has quite a bit of information outside of the hooks, and I have included a few additional properties in the `defaultBuildPrettyMeta` but this should probably be moved outside of the function as a more exhaustive list of safe properties is created

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
is this still draft?

**Comment by polymathca** [comment]
Ready to go!

File: fastify.js [hunk 1]
```diff
@@ -622,6 +636,12 @@ function fastify (options) {
     this[kErrorHandler] = func.bind(this)
     return this
   }
+
+  function printRoutes (opts = {}) {
+    // includeHooks:true - shortcut to include all supported hooks exported by fastify.Hooks
+    opts.includeMeta = opts.includeHooks ? opts.includeMeta ? supportedHooks.concat(opts.includeMeta) : supportedHooks : opts.includeMeta
```
Comment (Eomm) [inline fastify.js:642]: It is a little bit hard to read this statement and it is changing the input user object too
We could improve the readability

**Pull request merged**